libr/core/dmh_glibc.inc.c

```diff
@@ -99,6 +99,215 @@ static GHT GH(get_main_arena_with_symbol)(RCore *core, RDebugMap *map) {
 	return main_arena;
 }
 
+static GHT GH(get_section_size)(RCore *core, const char *path, const char *section_name) {
+	RBin *bin = core->bin;
+	RBinFile *bf = r_bin_cur (bin);
+	GHT size = GHT_MAX;
+
+	RBinFileOptions opt;
+	r_bin_file_options_init (&opt, -1, 0, 0, false);
+	if (!r_bin_open (bin, path, &opt)) {
+		R_LOG_ERROR ("get_section_size: r_bin_open failed on path %s", path);
+		return size;
+	}
+
+	RList *sections = r_bin_get_sections (bin);
+	RBinSection *section;
+	RListIter *iter;
+	r_list_foreach (sections, iter, section) {
+		if (!strcmp (section->name, section_name)) {
+			size = section->size;
+			break;
+		}
+	}
+
+	RBinFile *libc_bf = r_bin_cur (bin);
+	r_bin_file_delete (bin, libc_bf->id);
+	r_bin_file_set_cur_binfile (bin, bf);
+	return size;
+}
+
+static ut8 *GH(get_section_bytes)(RCore *core, const char *path, const char *section_name) {
+	RBin *bin = core->bin;
+	RBinFile *bf = r_bin_cur (bin);
+	bool found_section = false;
+	GHT paddr, size;
+	ut8 *buf = NULL;
+
+	RBinFileOptions opt;
+	r_bin_file_options_init (&opt, -1, 0, 0, false);
+	if (!r_bin_open (bin, path, &opt)) {
+		R_LOG_ERROR ("get_section_bytes: r_bin_open failed on path %s", path);
+		return NULL;
+	}
+
+	RBinFile *libc_bf = r_bin_cur (bin);
+	RList *sections = r_bin_get_sections (bin);
+	RBinSection *section;
+	RListIter *iter;
+	r_list_foreach (sections, iter, section) {
+		if (!strcmp (section->name, section_name)) {
+			found_section = true;
+			paddr = section->paddr;
+			size = section->size;
+			break;
+		}
+	}
+
+	if (!found_section) {
+		R_LOG_WARN ("get_section_bytes: section %s not found", section_name);
+		goto cleanup_exit;
+	}
+
+	// eprintf ("get_section_bytes: section found: %s size: %#08x  paddr: %#08x\n", section_name, size, paddr);
+
+	buf = calloc (size, 1);
+	if (!buf) {
+		R_LOG_ERROR ("get_section_bytes: calloc failed");
+		goto cleanup_exit;
+	}
+
+	st64 read_size = r_buf_read_at (libc_bf->buf, paddr, buf, size);
+
+	if (read_size != size) {
+		R_LOG_ERROR ("get_section_bytes: section read unexpected size: %#08x  (section->size: %d)", read_size, size);
+		free (buf);
+		buf = NULL;
+	}
+
+cleanup_exit:
+	r_bin_file_delete (bin, libc_bf->id);
+	r_bin_file_set_cur_binfile (bin, bf);
+	return buf;
+}
+
+R_API double GH(get_glibc_version)(RCore *core, const char *libc_path) {
+	double version = 0.0;
+
+	// First see if there is a "__libc_version" symbol
+	// If yes read version from there
+	GHT version_symbol = GH (get_va_symbol) (core, libc_path, "__libc_version");
+	if (version_symbol != GHT_MAX) {
+		FILE *libc_file = fopen (libc_path, "r ");
+		if (libc_file == NULL) {
+			R_LOG_WARN ("resolve_glibc_version: Failed to open %s", libc_path);
+			return false;
+		}
+		// TODO: futureproof this
+		char version_buffer[5];
+		fseek (libc_file, version_symbol, SEEK_SET);
+		fread (version_buffer, 1, 4, libc_file);
```

Comment:
the buffer is 5 byte length, but you fill only 4 bytes here. also check for return value.

Reply:
5th for null termination